shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -62,15 +62,52 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
       env: MutMap[Str, TypeInfo],
       mthEnv: MutMap[(Str, Str) \/ (Opt[Str], Str), MethodType],
       lvl: Int,
+      qenv: MutMap[Str, SkolemTag], // * SkolemTag for variables in quasiquotes
+      fvars: MutSet[ST], // * Free variables
+      inQuote: Bool, // * Is in quasiquote
       inPattern: Bool,
       tyDefs: Map[Str, TypeDef],
       tyDefs2: MutMap[Str, DelayedTypeInfo],
       inRecursiveDef: Opt[Var], // TODO rm
       extrCtx: ExtrCtx,
   ) {
-    def +=(b: Str -> TypeInfo): Unit = env += b
+    def +=(b: Str -> TypeInfo): Unit = {
+      env += b
+      // * For lambdas and let bindings, we should not have bindings with the same name in one context.
+      // * This check (`!qenv.contains(b._1)`) is for UCS desugared terms that contain bindings and it is safe to share the same skolem tag
+      // * since the temporary variables are always bound.
+      if (inQuote && !qenv.contains(b._1)) {
+        val tag = SkolemTag(freshVar(NoProv, N, nameHint = S(b._1))(lvl))(NoProv)
+        println(s"Create skolem tag $tag for ${b._2} in quasiquote.")
+        qenv += b._1 -> tag
+      }
+    }
     def ++=(bs: IterableOnce[Str -> TypeInfo]): Unit = bs.iterator.foreach(+=)
-    def get(name: Str): Opt[TypeInfo] = env.get(name) orElse parent.dlof(_.get(name))(N)
+    def get(name: Str): Opt[TypeInfo] = {
+      // * 1. If we try to get a quoted variable from a quoted context, or get a non-quoted variable from a non-quoted context,
+      // * we can directly search the `env`;
+      // * 2. If we try to get a quoted variable from a non-quoted context, we need to wrap it with its contextual Skolem;
+      // * 3. Otherwise, it is a top-level defined symbol, or the context has a quoted parent.
+      def rec(ctx: Ctx): Opt[TypeInfo] = {
+        if (inQuote === ctx.inQuote) ctx.env.get(name) orElse ctx.parent.flatMap(rec(_))
+        else if (!inQuote) (ctx.env.get(name), ctx.qget(name)) match {
+          case (S(VarSymbol(ty, _)), S(tag)) =>
+            S(VarSymbol(TypeRef(TypeName("Var"), ty :: tag :: Nil)(noProv), Var(name)))
+          case _ => ctx.parent.flatMap(rec(_))
+        }
+        else ctx.parent match {
+          case S(parent) => rec(parent)
+          case _ => ctx.env.get(name)
+        }
+      }
+      rec(this)
+    }
+    def qget(name: Str): Opt[SkolemTag] = qenv.get(name) orElse parent.dlof(_.qget(name))(N)
```

Comment:
Rename.

Reply:
I've updated all names and comments mentioned above.